### PR TITLE
Skip trivial Resize, handle resize of Broadcast at concretization.

### LINKED ISSUE #798: Internal assert for pad: Failed to concretize an output IterType for expression
To reproduce:
```py
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[1, -1], contiguity=[None, True], dtype=DataType.Int, is_cpu=False)
    S1 = fd.define_scalar(None, dtype=DataType.Int)
    T8 = fd.ops.pad(T0, [0, 0, 0, 0], S1)
    fd.add_output(T8)

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

inputs = [
    torch.randint(0, 10, (1, 3), dtype=torch.int64, device='cuda:0').as_strided((1, 3), (3, 1)),
    -5,
]
fd.execute(inputs)
```
```py
File ~/dev/pytorch/master/nvfuser/__init__.py:122, in FusionDefinition.execute(self, inputs, device, override_user_schedule, capture_debug_output)
    120 result = None
    121 try:
--> 122     result = self._execute(
    123         inputs,
    124         override_user_schedule,
    125         device=device,
    126         capture_debug_output=capture_debug_output,
    127     )
    128 except Exception as err:
    129     msg = (
    130         f"An error occurred while executing nvFuser FusionDefinition {self.id()}.\n"
    131         "If you believe this is a bug or need assistance, please file an issue at "
    132         "https://github.com/NVIDIA/Fuser/issues/new\n"
    133     )

RuntimeError: iter_type != IterType::Symbolic INTERNAL ASSERT FAILED at "/home/iyashchuk/dev/nvfuser/csrc/dynamic_transform.cpp":634, please report a bug to PyTorch. Failed to concretize an output IterType for expression: Resize: bS2{1}rf by ( (nvfuser_index_t)(0) ) and ( (nvfuser_index_t)(0) ) -> bS3{( ( ( (nvfuser_index_t)(0) ) + 1 ) + ( (nvfuser_index_t)(0) ) )}rf
```

### PR DESCRIPTION
**Pull request by jacobhinkle**
This updates the root to rfactor propagation in IterType concretization of dynamic fusions.

Previously, although we only overwrote Symbolic IterDomains in this step, we still asserted that we could infer an IterType for each I moved that check so that it is only applied when we need to make a change.

Additionally, we previously propagated Broadcast-only IterDomains as Symbolic, since we combine with our previous estimate using promoteIterType. As mentioned in a comment, this means Broadcast gets propagated as Symbolic. Instead we now only fall back to promoteIterType when there are multiple input IterTypes to the IterDomain expression.

Fixes #798

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

File: csrc/ir/nodes.cpp [hunk 1]
```diff
@@ -2823,12 +2823,16 @@ IterDomain* IterDomain::resize(
   if (iter_type_opt.has_value()) {
     iter_type = iter_type_opt.value();
   } else if (left_expansion->isConstInt() && right_expansion->isConstInt()) {
-    if (resized_id_size->isConstInt()) {
+    auto left = left_expansion->evaluateInt();
+    auto right = right_expansion->evaluateInt();
+    if (left == 0 && right == 0) {
+      // Trivial Resize
+      return in;
```
Comment (naoyam) [inline csrc/ir/nodes.cpp:2830]: It seems odd to check the trivial case here as this part is to figure out the iteration type. Why not just do this check early? Also, maybe we should check the optional iteration type is not given or the same as the input iteration type.

File: csrc/dynamic_transform.cpp [hunk 2]
```diff
@@ -629,18 +629,29 @@ void DynamicTransformConcretizer::mutate(TensorView* tv) {
...
       for (auto inp_id : ir_utils::filterByType<IterDomain>(expr->inputs())) {
         auto updated_id = maybeMutated(inp_id)->as<IterDomain>();
-        iter_type = ops::promoteIterType(iter_type, updated_id->getIterType());
+        // ops::promoteIterType will favor Symbolic if it encounters it
+        // alongside Broadcast. This is preferable at fusion definition, but
+        // here we are propagating, and if we only see Broadcast in some
+        // dimension, then we should not retain Symbolic. To work around this,
+        // we always overwrite Symbolic with the first concrete IterType we
+        // encounter.
+        iter_type = iter_type == IterType::Symbolic
```
Comment (naoyam) [inline csrc/dynamic_transform.cpp:638]: Here, `updated_id->getIterType()` should be always non symbolic, right?

Comment (naoyam) [reply csrc/dynamic_transform.cpp:638]: This feels more intuitive to me.

```
for (const auto i: c10::irange(...)) {
    auto updated_id = ...;
  TORCH_INTERNAL_ASSERT(updated_id->getIterType() != IterType::Symbolic);
  if (i == 0) {
    iter_type = updated_id->getIterType();
  } else {
    iter_type = ops::promoteIterType(iter_type, updated_id->getIterType());
  }
}
```

File: csrc/ir/nodes.cpp (hunk 1 above)
Comment (jacobhinkle) [reply csrc/ir/nodes.cpp]: Good point. I moved the check to the beginning of the function and added an assertion that the specified `IterType` does not clash, if it is given.

File: csrc/dynamic_transform.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/dynamic_transform.cpp:638]: Could you give me an example?

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the update!

**Pull request merged**